Enable errorprone globally

iemejia opened the pull request:
R: @swegner

swegner commented on build_rules.gradle:
```diff
@@ -560,6 +560,10 @@ ext.applyJavaNature = {
     tasks.withType(JavaCompile) {
       options.compilerArgs += "-XepDisableWarningsInGeneratedCode"
     }
+    dependencies {
```
Should this be outside the `enableErrorProne` block? My understanding is that these annotations are for Findbugs.

iemejia replied:
Well these imports were mostly added to remove warnings on module-info.java files reported by error-prone. I can relocate it if you prefer but I have the impression that this scope is tighter.

swegner replied:
I don't have a strong preference. Maybe add a comment explaining why it's here since it's not totally obvious why ErrorProne usage should require a FindBugs dependency.

iemejia replied:
ok

swegner commented on sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/util/RetryHttpRequestInitializerTest.java:
```diff
@@ -255,6 +255,7 @@ public Integer answer(InvocationOnMock invocation) {
    * is invoked.
    */
   @Test
+  @SuppressWarnings("AssertionFailureIgnored")
```
Can this test be improved such that we don't need this suppression? The [ErrorProne AssertionFailureIgnored](https://errorprone.info/bugpattern/AssertionFailureIgnored) page suggests using JUnit's `assertThrows()`.

iemejia replied:
I did not find an easy way to do this and also the proposed solution by errorprone is invalid. See https://github.com/google/error-prone/issues/1059
Any suggestions are welcome.

swegner replied:
You could re-create the idea behind `assertThrows` with:

```java
Throwable failure = null;
try {
  getRequest.execute();
} catch (Throwable e) {
  failure = e;
}
assertNotNull("Expected execute to throw an exception", failure);
assertThat(failure, Matchers.instanceOf(SocketTimeoutException.class));
assertEquals(1 + defaultNumberOfRetries, executeCount.get());
expectedLogs.verifyWarn("performed 10 retries due to IOExceptions");
```

iemejia replied:
I don't know why I was thinking in a way more complex idea. Thanks, added this.

swegner commented on build_rules.gradle:
```diff
@@ -344,7 +344,7 @@ class JavaNatureConfiguration {
   /** Controls whether the errorprone plugin is enabled and configured */
   boolean enableErrorProne = true
   /** Controls whether compiler warnings are treated as errors */
-  boolean failOnWarning = false
```
My preference is to remove this configuration altogether and make the `failOnWarning` behavior default.

iemejia replied:
We can't remove this for the moment we have modules still skipping using gthis e.g. 'sdks/java/core'.

swegner replied:
Got it. I'll work on finishing up `sdks/java/core` as I have time and then remove this. /cc @pabloem

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
Addressed all comments, rebased after the plugin changes, and merged, only missing thing now is 'sdks/java/core'. Probably it would be a good idea to submit an email when finished @swegner to notify and thank the contributors to this.

swegner commented:
Makes sense to me-- I can send a notification once we're fully-migrated. I probably won't get to the remaining `sdks-java-core` work until next week. If anybody has time before that, feel free to pick it up.
